### DISCUSSION TITLE
Release 0.3: Complex numbers

### DISCUSSION THREAD

**Discussion by c3d** [c3d/db48x#133] [Announcements]
Bug fixes:

- CHS on second number on the command line (#105)
- program interruption when keys were pressed other than EXIT (#104)
- Rounding of 1.9999 displaying as 10.000 instead of 2.000 (#112)
- Memory corruption in logical operations on big numbers (#103, #107)
- Out of memory error when no spacing is present (#109)
- Strip trailing zeros introduced by rounding (#116, #117)
- Reject `0 SIG` (displaying no significant digits)
- Potential crash in low-memory condition (#118)
- Removed useless "Program" menu in system menu (#120)
- Exit the state save/load menu after saving/loading (#122)
- Crash loading state file if no error was generated
- Detection of integer overflow in implementation of a^b (power)
- Evaluation of `"ABC" 3 *` was no longer working

Features:
- Complex numbers (#15)
- Render based numbers with base as subscript (#106)
- Change default spacing for fractional part to 5 digits (#108)
- Animation for long-running user programs and garbage collection (#115)
- xshift-EXIT saves state to disk (#113)
- Switch to three-level menus by default (#128)
- Generalization of `pow` with integer on right side
- Represent `hypot` and `atan2` using ⊿ and ∠ operators
- Added `sign` function
- Change behavior of Shift-Enter (alpha), now cycles back to non-alpha

Other:
- Add keyboard cutout for tentative keyboard layout
- Vastly enhanced regression test suite (faster and more extensive)
- Fixed dependencies in Makefile for fonts
- Improved efficiency of arithmetic computations
- Added several characters to the base font
- Improve 'const'-safety in C++ code, garbage-collected pointers are const now

Signed-off-by: Christophe de Dinechin <christophe@dinechin.org>

**Full Changelog**: https://github.com/c3d/DB48X-on-DM42/compare/v0.2.4...v0.3

<hr /><em>This discussion was created from the release <a href='https://github.com/c3d/DB48X-on-DM42/releases/tag/v0.3'>Release 0.3: Complex numbers</a>.</em>

**Comment by balta1980** [comment]
Hi, thank you for the new release. After some usage, I will give you some feedback:

-When you use the E exponential button and you enter a number example 1E3, it returns "1000." instead of "1,000." Missing the thousands separator And if you multiply that number by another it still is missing the thousands separator. But if you add or rest a number it gets back its thousands separator.

- Only on battery, when the display is showing a menu, input numbers and operating with them is a little slow. There is a delay, but when you press exit and close menus everything goes faster again.

Again, thank you, great job.

**Reply by c3d** [reply]
> -When you use the E exponential button and you enter a number example 1E3, it returns "1000." instead of "1,000." Missing the thousands separator And if you multiply that number by another it still is missing the thousands separator. But if you add or rest a number it gets back its thousands separator.

This looks very similar to #137.

> Only on battery, when the display is showing a menu, input numbers and operating with them is a little slow. There is a delay, but when you press exit and close menus everything goes faster again.

I thought that I had fixed that, but apparently not. I will look into it. I opened #153 to track it.

**Reply by c3d** [reply]
@balta1980 I just posted v0.3.1, which has, among other things, a significant acceleration of the text editor. #137 is still not fixed, I thought it was less important.

**Reply by balta1980** [reply]
Now it is fixed, no lags. Thank you.
